### PR TITLE
Skip unnecessary directory iteration

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today `NodeEnvironment#findAllShardIds` enumerates the index directories
in each data path in order to find one with a specific name. Since we
already know the name of the folder we seek we can construct the path
directly and avoid this directory listing. This commit does that.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/env/NodeEnvironment.java [hunk 1]
```diff
@@ -1044,15 +1044,8 @@ public String nodeId() {
...
-            Path location = nodePath.indicesPath;
-            if (Files.isDirectory(location)) {
-                try (DirectoryStream<Path> indexStream = Files.newDirectoryStream(location)) {
-                    for (Path indexPath : indexStream) {
-                        if (indexUniquePathId.equals(indexPath.getFileName().toString())) {
-                            shardIds.addAll(findAllShardsForIndex(indexPath, index));
-                        }
-                    }
-                }
+            if (Files.isDirectory(nodePath.indicesPath)) {
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/env/NodeEnvironment.java:1047]: Shouldn't this be checking `nodePath.indicesPath.resolve(indexUniquePathId)`?

**Review by rjernst — COMMENTED** [review]
Looks good, one question

File: server/src/main/java/org/elasticsearch/env/NodeEnvironment.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/env/NodeEnvironment.java:1047]: This preserves the check in the original code; we already check `Files.isDirectory(nodePath.indicesPath.resolve(indexUniquePathId))` within `findAllShardsForIndex`. Maybe we should just remove this check entirely?

**Review by rjernst — APPROVED** [review]
LGTM

**Pull request merged**